src/compiler/build/build-stats.ts

```diff
@@ -0,0 +1,177 @@
+import { byteSize, sortBy } from '@utils';
+import type * as d from '../../declarations';
+
+/**
+ *
+ * @param config
+ * @param compilerCtx
+ * @param buildCtx
+ * @returns
+ */
+export function generateBuildStats(
+  config: d.Config,
+  buildCtx: d.BuildCtx
+): d.CompilerBuildStats | { diagnostics: d.Diagnostic[] } {
+  const buildResults = buildCtx.buildResults;
+
+  let jsonData: d.CompilerBuildStats | { diagnostics: d.Diagnostic[] };
+
+  try {
+    if (buildResults.hasError) {
+      jsonData = {
+        diagnostics: buildResults.diagnostics,
+      };
+    } else {
+      const stats: d.CompilerBuildStats = {
+        timestamp: buildResults.timestamp,
+        compiler: {
+          name: config.sys.name,
+          version: config.sys.version,
+        },
+        app: {
+          namespace: config.namespace,
+          fsNamespace: config.fsNamespace,
+          components: Object.keys(buildResults.componentGraph).length,
+          entries: Object.keys(buildResults.componentGraph).length,
+          bundles: buildResults.outputs.reduce((total, en) => (total += en.files.length), 0),
+          outputs: getAppOutputs(config, buildResults),
+        },
+        options: {
+          minifyJs: config.minifyJs,
+          minifyCss: config.minifyCss,
+          hashFileNames: config.hashFileNames,
+          hashedFileNameLength: config.hashedFileNameLength,
+          buildEs5: config.buildEs5,
+        },
+        formats: {
+          esmBrowser: sanitizeBundlesForStats(buildCtx.esmBrowserComponentBundle),
+          esm: sanitizeBundlesForStats(buildCtx.esmComponentBundle),
+          es5: sanitizeBundlesForStats(buildCtx.es5ComponentBundle),
+          system: sanitizeBundlesForStats(buildCtx.systemComponentBundle),
+          commonjs: sanitizeBundlesForStats(buildCtx.commonJsComponentBundle),
+        },
+        components: getComponentsFileMap(config, buildCtx),
+        entries: buildCtx.entryModules,
+        componentGraph: buildResults.componentGraph,
+        sourceGraph: getSourceGraph(config, buildCtx),
+        rollupResults: buildCtx.rollupResults,
+        collections: getCollections(config, buildCtx),
+      };
+
+      jsonData = stats;
+    }
+  } catch (e) {
+    jsonData = {
+      diagnostics: [e.message],
+    };
+  }
+
+  return jsonData;
+}
+
+/**
+ *
+ * @param config
+ * @param buildCtx
+ * @returns
+ */
+export async function writeBuildStats(config: d.Config, data: d.CompilerBuildStats | { diagnostics: d.Diagnostic[] }) {
+  const statsTargets = config.outputTargets.filter((o) => o.type === 'stats') as d.OutputTargetStats[];
```

Comment:
Normally when we filter output targets, we do something like
```
  const outputTargets = config.outputTargets.filter(isOutputTargetDistCustomElementsBundle);
```
where `isOutputTargetDisCustomElementsBundle` is defined in [src/compiler/output-targets/output-utils.ts](https://github.com/ionic-team/stencil/blob/master/src/compiler/output-targets/output-utils.ts) as
```
export const isOutputTargetDistCustomElementsBundle = (
  o: d.OutputTarget
): o is d.OutputTargetDistCustomElementsBundle => o.type === DIST_CUSTOM_ELEMENTS_BUNDLE;
```
and `DIST_CUSTOM_ELEMENTS_BUNDLE` is a constant we define in that same file.  Is there a reason we don't do that here?

I think most of the machinery is already in place [here](https://github.com/ionic-team/stencil/blob/9fc76579a3a3e1127ba43b354a572ac40eda3770/src/compiler/output-targets/output-utils.ts#L95)

```suggestion
  const statsTargets = config.outputTargets.filter(isOutputTargetStats);
```
(you'll need to import `isOutputTargetStats` though)

Reply:
TIL! Thank you!